### ISSUE TITLE
GoPro HERO10 Black Raw file has no white balance 

### ISSUE THREAD

**Issue by supertobi** [darktable-org/darktable#10072]
**To Reproduce**

1. open the attached file (I've added the file to raw.pixls.us too)
[GOPR0119.zip](https://github.com/darktable-org/darktable/files/7226151/GOPR0119.zip)
2. unzipp it 
3. open it in dt
4. see error

**Expected behavior**
less pink

**Screenshots**
![grafik](https://user-images.githubusercontent.com/653976/134687303-268b1103-d593-40b4-a857-3fafedf7a742.png)

_A bisect is much appreciated and can significantly simplify the developer's job._
_HowTo: https://github.com/darktable-org/darktable/wiki#finding-bug-causes and https://www.youtube.com/watch?v=D7JJnLFOn4A_

**Platform**
* darktable version : 3.6.0
* OS                : Win10 Pro (Build 19043.1237)
* Memory            : 16GB
* Graphics card     : AMD Radeon 540X Series

**Comment by kmilos** [comment]
There might be something fishy going on with the HERO10 `CFAPattern`, `BlackLevel`, `WhiteLevel`, and `AsShotNeutral` tags

```
Exif.Image.CFAPattern                         1 2 0 1
Exif.Image.BlackLevel                         51200/256 51200/256 51200/256 51200/256
Exif.Image.WhiteLevel                         4095
Exif.Image.AsShotNeutral                      0/1000000 0/1000000 0/1000000
```

when compared to e.g. a HERO9 sample

```
Exif.Image.CFAPattern                         0 1 1 2
Exif.Image.BlackLevel                         0/256 0/256 0/256 0/256
Exif.Image.WhiteLevel                         16383
Exif.Image.AsShotNeutral                      487619/1000000 1000000/1000000 543524/1000000
```

Sensor is supposedly the same going by the reviews and press releases so I would expect the same `CFAPattern`, `BlackLevel`, and `WhiteLevel`....

`WhiteLevel` is furthermore used as a parameter for rawspeed's VC5 decompressor so that one seems extra important... 

0,0,0 white balance seems just plain wrong...

Does it open ok in any other apps?

**Comment by supertobi** [comment]
> Does it open ok in any other apps?

I've just tested RawTherapee, but RawTherapee doesn't support GPR files.

Adobe DNG Converter is able to convert the file and creates this DNG:
[GOPR0119.dng.zip](https://github.com/darktable-org/darktable/files/7235244/GOPR0119.dng.zip)
 
I haven't tested the free  GPR converter from GoPro:
https://github.com/gopro/gpr

**Comment by kmilos** [comment]
Interesting, same `CFAPattern`, `BlackLevel`, and `WhiteLevel` are preserved in the DNG and it indeed works... Both files complain about missing white balance information and seem to fall back to some D65 default, so don't think that one was crucial...

It is possible then that there is indeed something wrong going in in the rawspeed decoder/decompressor... There's been some changes around VC5 as I've noticed recently, do you mind trying out [the current development build](https://discuss.pixls.us/t/darktable-windows-insider-program-9-26/27121)?

**Comment by kmilos** [comment]
I tested 3.7.0+1122_g362a952f6, no change...

At this point I think it's best to close this issue, and [open a formal camera support](https://github.com/darktable-org/darktable/wiki/Camera-support) one (link to this one in there though).

**Comment by supertobi** [comment]
I think we don't need a new issue. Adding the tag "scope: camera support" and assigning it to @LebedevRI does the trick.

**Comment by LebedevRI** [comment]
Can you contribute a sample to RPU that has been directly copied from camera and not modified in any way?

**Comment by LebedevRI** [comment]
I guess let's wait on https://github.com/gopro/gpr/issues/33 first?

**Comment by supertobi** [comment]
@LebedevRI 
1. The file I added here is directly copied from the camera and not modified in any way and I already uploaded it to RPU.
2. Do be honest, I don't expect any reaction on gopro/gpr#33

**Comment by LebedevRI** [comment]
> * The file I added here is directly copied from the camera and not modified in any way and I already uploaded it to RPU.

I see. It looked like it had some windows filepaths, but then the other files also have that.
Honestly this bug is rather sad, because i'm not quite sure which particular wording in DNG spec
says that we should detect that WB (all-zeros) as invalid and use camera-neutral one instead.

**Comment by kmilos** [comment]
I don't think it's a white balance issue - in both the GPR and converted DNG case the field is invalid and dt gives up and defaults to something sane (D65). It looks more like a decoding issue to me: no matter how you change the black & white levels in dt and white balance, the image looks pink...

Here's the histogram of the GPR once I set the black level to 0 and white level to 16384 (white balance is at 6504K):

![image](https://user-images.githubusercontent.com/4973094/135056361-6e9e7c46-b15b-4a1f-9df4-386244dfad50.png)

Here's the histogram of the Adobe's VC5 decode -> JPEG reencode from the DNG (same black/white and WB settings):

![image](https://user-images.githubusercontent.com/4973094/135056884-509249a4-0553-4ca4-b921-f6ce39e3465d.png)

It's as if like the color planes are swapped almost, but the decoded values are also somehow strange (esp. for the green plane)...

**Comment by supertobi** [comment]
Perhaps its a different pattern of the sensor, have a look at the exif information:

HERO10
> Exif.Image.CFAPattern                         1 2 0 1

when compared to a HERO9 sample

> Exif.Image.CFAPattern                         0 1 1 2

**Comment by LebedevRI** [comment]
Yeah, it's pretty obvious [wrong](https://github.com/darktable-org/rawspeed/blob/fb42fcdb4c296006146d548226453b6d029ccfc0/src/librawspeed/decompressors/VC5Decompressor.cpp#L847-L850), just need to unbreak the compiler first..

**Comment by kmilos** [comment]
It seems more than that to me, the decoded values are really strange - lots of them come out at lower than the supposed black level of 200 in that green channel...

**Comment by kmilos** [comment]
Confirmed that indeed

```C++
        out(2 * row + 0, 2 * col + 0) = static_cast<uint16_t>(mVC5LogTable[g1]);
        out(2 * row + 0, 2 * col + 1) = static_cast<uint16_t>(mVC5LogTable[b]);
        out(2 * row + 1, 2 * col + 0) = static_cast<uint16_t>(mVC5LogTable[r]);
        out(2 * row + 1, 2 * col + 1) = static_cast<uint16_t>(mVC5LogTable[g2]);

```

Solves the issue. I used something like `bool rggb = mRaw->cfa.getDcrawFilter() == 0x94949494;` as the condition...

**Comment by LebedevRI** [comment]
Heh.

**Comment by supertobi** [comment]
> Yeah, it's pretty obvious [wrong](https://github.com/darktable-org/rawspeed/blob/fb42fcdb4c296006146d548226453b6d029ccfc0/src/librawspeed/decompressors/VC5Decompressor.cpp#L847-L850), just need to unbreak the compiler first..

How is your compiler doing?

**Comment by arbasel** [comment]
Hi, I see that this issue is still happening. Is there some way for me to fix it?

Thanks

**Comment by supertobi** [comment]
@arbasel the easiest you can do is to use @kmilos patch. It's not elegant but works. Or provide a better patch, that satisfies @LebedevRI .

**Comment by jazuntee** [comment]
> @arbasel the easiest you can do is to use @kmilos patch. It's not elegant but works. Or provide a better patch, that satisfies @LebedevRI .

How do use @kmilos patch? Is that something I can plug-in to the current version of the app or is that something that requires recompiling?

**Comment by supertobi** [comment]
It requires recompiling. And I think it should be merged, even if it is ugly. It works and is better then the something that doesn't wok.

**Comment by LebedevRI** [comment]
FWIW, the other problem (at least on the HERO10 samples available on RPU), is that they simply lack white balance in EXIF.

**Comment by kmilos** [comment]
I think this can be closed now as Roman merged a proper fix for 4.2? @dterrahe @LebedevRI

**Comment by LebedevRI** [comment]
Technically, no. HERO10 indeed do not have camera white balance, at least the samples on RPU.
